### ISSUE TITLE
Training a pruned model makes it 10 times slower at inference 

### ISSUE THREAD

**Issue by Coderx7** [pytorch/pytorch#46180]
## 🐛 Bug

I pruned a resnet18 model and saved it as jit so it can be used in libtorch. The model is pruned and trained using Pytorch 1.5.1 and Python 3.7 under linux. Everything seems fine, the model learns and performs as expected.
However, what is not normal is the 10x times slow-down at inference. 
Below are the stats I recorded for 10 iterations of the pruned model's forward time before and after being trained:   
The following times are calculated in libtorch 1.6cpu under windows. 
```
Before training : 7,405 ms  
After training :  75,480 ms
```
Both models (un-trained pruned model + the trained one) : https://gofile.io/d/sFkJkO 
## To Reproduce

1. grab a model, prune it (you can use [torch_pruning](https://github.com/VainF/Torch-Pruning) )
2. train it
3. when saving the checkpoint, save it as jit model

## Expected behavior

There shouldn't be any speed difference between the two!

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
Collecting environment information...
PyTorch version: 1.6.0+cpu
Is debug build: No
CUDA used to build PyTorch: None

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
CMake version: version 3.17.2

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] facenet-pytorch==2.2.9
[pip] numpy==1.18.1
[pip] numpydoc==0.9.1
[pip] torch==1.6.0+cpu
[pip] torch-pruning==0.2.2
[pip] torchfile==0.1.0
[pip] torchnet==0.0.4
[pip] torchprof==1.0.0
[pip] torchscope==0.1.0
[pip] torchstat==0.0.7
[pip] torchvision==0.7.0+cpu
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] facenet-pytorch           2.2.9                    pypi_0    pypi
[conda] mkl                       2019.0                   pypi_0    pypi
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.0.15           py37h14836fe_0
[conda] mkl_random                1.1.1            py37h47e9c7a_0
[conda] numpy                     1.18.1           py37h93ca92e_0
[conda] numpy-base                1.18.1           py37hc3f5095_1
[conda] numpydoc                  0.9.1                      py_0
[conda] torch                     1.6.0+cpu                pypi_0    pypi
[conda] torch-pruning             0.2.2                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchnet                  0.0.4                    pypi_0    pypi
[conda] torchprof                 1.0.0                    pypi_0    pypi
[conda] torchscope                0.1.0                    pypi_0    pypi
[conda] torchstat                 0.0.7                    pypi_0    pypi
[conda] torchvision               0.5.0+cpu                pypi_0    pypi
PS C:\Users\User>
```

cc @gmagogsfm

**Comment by glaringlee** [comment]
@Coderx7 
If I understand correctly. 
Can you use 'trained parameters directly into your module and run inference'  compare with 'train module first and run inference after'? Let me know whether both inference have the same performance?

The thing is that after training, your module's parameter changed, different inputs do affect the performance of the pytorch operations sometimes (eg, some operators might go different code path based on different shapes, strides, etc).

And, it is not feasible to debug a 50+ MB module at out end, can you simplify your module or adding more timing log to locate which operator is slow?

**Comment by Coderx7** [comment]
Hi, 
I couldnt underastand this part: 

> Can you use 'trained parameters directly into your module and run inference' compare with 'train module first and run inference after'?

Do you mean to transfer the trained parameters from a trained model to the pruned version thats not trained yet? and see if the performance is any different?

I need to add that both networks use the very same image-size both at training and inference. this shouldn't be affecting this much of a slow down anyway right? 

Yes, you are right, I need to be providing an MRE (minimal repreducible example), but I  thought maybe by just looking at the models generated ILs in jit, you could comeup with some theory at least.

Also the models are jit traced how am I suppose  to time each operator? the time that I posted, are simply the take it takes for the `forward()` to does its job.
Do you mean to put different timmings in model definition (python model) and print the timmings as the forward happens? 
will this be transfered and usabe when jit traced?

**Comment by glaringlee** [comment]
@Coderx7 
Yep, that's what I meant, but now I see what you mean. So you train your module under python and save it with jit, and run inference with libtorch using jit saved module, right? If so, please discard my previous comments, I've thought you were running in the same env.

Another thing I want to ask is that did you set the same thread number when doing training and inference?
You can try set the thread number to one in python when doing the training, see what is the running time. 
Add this line in your python code before doing the training `torch.set_num_threads` , let me know whether there is any difference.

**Comment by Coderx7** [comment]
Yes you are right. 
No everything thread related is default. I didnt change anything. 
alright I'll come back to you when this is all finished and ready.

**Comment by Coderx7** [comment]
OK here are the results. the `thread_num` is not set to 1 anywhere, since when I set it the performance gets several times worse. this is shown below : 
![pytorch_pruning_bench](https://user-images.githubusercontent.com/5382892/95818075-df8b3b80-0d2f-11eb-9695-86c0bf3cf797.gif)
As the name suggests, one of the models are fully trained after being pruned, and the other is just prior to being pruned. I also tested with the model thats trained for only one epoch and it acts normally (i.e. just like the model being freshly pruned and not t rained! again that is it runs fast)  

I also tested this under Linux and Pytorch 1.5.1, Windows with Pytorch 1.6.0 and libtorch 1.6.0 and the result in all of them is the same.  So hopefully this is not some windows specific issue!
I also tested with GraphOptimization on the libtorch side and seems `GraphExecutorOptimize` is false by default since setting it to true will again make thing worse. 

Is there anything else that you would like to know? before I delve into creating an MRE that replicates this?

**Comment by glaringlee** [comment]
@Coderx7 
I see. You actually jit saved both models and run under C++. It seems to me that the timing difference do caused by model itself since jit saving process is always the same (to me), shouldn't cause any difference. I will have jit people take a look at this.

**Comment by glaringlee** [comment]
Send to jit oncall to take a look at, please advise.

**Comment by wconstab** [comment]
@Coderx7 thanks for offering to produce a minimal repro- that will help us a lot.  If you can share the script for the simple model and the script that saves/loads it, that would be best.

**Comment by wconstab** [comment]
@Coderx7 Just wanted to follow up- if you'll be able to produce a simpler example for us to debug that'd help a lot!

**Comment by Coderx7** [comment]
@wconstab I'm on it! I have been training on a new model, the simplest model I could. I noticed during training there is no slow down whatso ever. so far. I'll update tomorrow when the training is finished. I'll see what changes we have

**Comment by Coderx7** [comment]
Here's a small update of the ongoing process. 
I've tried to replicate this issue with cifar10 classification to no avail.
I also tried the arcface loss which I'm using at work again to no avail thinking maybe somehow this has something to do with it! Eveything works fine on the cifar10 MRE(minimal repreducible example) that I have been working on so far. 
However, I have not yet tried the DataParallel so far, I'm not sure if this plays a role or not. I also am not sure if the cifar10 requires more training to show the symptoms as in my previous findings, in the early stages of the training the models do not show any decrease in performance when jit traced! 
I have been using a simple architecture and this might have something to do with it! I'm not sure. I'll be testing the resnet again on cifar10 as well.   
I've got some more tests to run and I'll update how this goes.   
Honestly this is very bizarre, anything that could kind of help me into directing my tests would be greatly appreciated. any kind of intuitions, anything  or else I'm going to just test every single scenario I can think of !

**Comment by wconstab** [comment]
just off the top of my head, i guess a couple things you could check quickly to narrow it down are
* threading: i didn't fully understand what your threading setup is, but you can run the fast and slow with htop open and get a visual sense of whether there are a different number of threads being used during the compute intensive part.  if so, it may be an issue with one of the _NUM_THREADS variables that controls openmp, mkl, etc.

* which kernels/ops are being called - you can print out jit_model.code and jit_model.graph (in python; c++ would be a different api) and compare the dump for the fast and slow model.  the .graph would show a lot of detail, but if you side-by-side diff you might be able to see whether a different gemm/conv operator is being used or if there is anything else obvious.

**Comment by gmagogsfm** [comment]
Closing due to inactivity. Please reopen if there are new findings or remaining issues.
